src/core/stdcpp/allocator.d

```diff
@@ -0,0 +1,66 @@
+/**
+ * D header file for interaction with C++ std::allocator.
+ *
+ * Copyright: Copyright (c) 2019 D Language Foundation
+ * License: Distributed under the
+ *      $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost Software License 1.0).
+ *    (See accompanying file LICENSE)
+ * Authors:   Manu Evans
+ * Source:    $(DRUNTIMESRC core/stdcpp/allocator.d)
+ */
+
+module core.stdcpp.allocator;
+
+// some versions of VS require a `* const` pointer mangling hack
+// we need a way to supply the target VS version to the compile
+version (CppRuntime_Microsoft)
+    version = NeedsMangleHack;
+
+extern(C++, "std"):
+
+/**
+ * Allocators are classes that define memory models to be used by some parts of
+ * the C++ Standard Library, and most specifically, by STL containers.
+ */
+extern(C++, class) struct allocator(T)
+{
+    static assert(!is(T == const), "The C++ Standard forbids containers of const elements because allocator!(const T) is ill-formed.");
+    static assert(!is(T == immutable), "immutable is not representable in C++");
+
+    alias value_type = T;
+
+    version (CppRuntime_Microsoft)
+    {
+        T* allocate(size_t count) nothrow @nogc;
+        void deallocate(T* ptr, size_t count) nothrow @nogc;
+
+        version (NeedsMangleHack)
+        {
+            // HACK: workaround to make `deallocate` link as a `T * const`
+            extern (D) private static string constHack(string name)
+            {
+                version (Win64)
+                    enum sub = "AAXPE";
+                else
+                    enum sub = "AEXPA";
+                foreach (i; 0 .. name.length - sub.length)
+                    if (name[i .. i + sub.length] == sub[])
+                        return name[0 .. i + 3] ~ 'Q' ~ name[i + 4 .. $];
+                assert(false, "substitution string not found!");
+            }
+            pragma(linkerDirective, "/alternatename:" ~ deallocate.mangleof ~ "=" ~ constHack(deallocate.mangleof));
+        }
+    }
+    else
+    {
+        T* allocate(size_t count, const(void)* = null) nothrow @nogc;
+        void deallocate(T* ptr, size_t count) nothrow @nogc;
+    }
+
+    // TODO: should we have D helpers like these?
```

Comment:
Do you expect there will be a lot of allocations from the D side?

Reply:
hard to know. If someone is linking against C++, and they're doing their memory management in C++, then I can imagine that the D code might also want to use the same allocator locally. Perhaps it would be wrapped by a `std.experimental.allocator`, but also perhaps not...?

Reply:
How about we wait with adding these helpers to see if there's a need? It's easier to add symbols than to remove.

Reply:
Well that's why they're commented out ;)